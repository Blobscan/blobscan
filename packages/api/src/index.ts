import type { inferRouterInputs, inferRouterOutputs } from "@trpc/server";

import type { AppRouter } from "./app-router";

export { appRouter, type AppRouter } from "./app-router";
export { createTRPCContext, createTRPCInnerContext } from "./context";
export type {
  TRPCInnerContext as TRCInnerContext,
  TRPCContext,
} from "./context";

export * from "@trpc/server/adapters/express";

export { env } from "./env";
export { gracefulShutdown } from "./graceful-shutdown";
export { metricsHandler } from "./instrumentation";
export type {
  BlobStorage,
  Rollup,
  Metric,
  MetricHistogram,
  MetricHistogramBucket,
  Metrics,
  MetricsClient,
} from "@blobscan/db";

/**
 * Inference helpers for input types
 * @example type HelloInput = RouterInputs['example']['hello']
 **/
export type RouterInputs = inferRouterInputs<AppRouter>;

/**
 * Inference helpers for output types
 * @example type HelloOutput = RouterOutputs['example']['hello']
 **/
export type RouterOutputs = inferRouterOutputs<AppRouter>;
