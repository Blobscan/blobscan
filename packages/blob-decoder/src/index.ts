export * from "./blob-decoder";
export * from "./types";
