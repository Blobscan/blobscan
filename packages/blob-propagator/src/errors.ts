type ErrorCause = Error | string;

export class ErrorException extends Error {
  constructor(message: string, cause?: ErrorCause) {
    super(message, {
      cause,
    });

    this.name = this.constructor.name;
  }
}

export class BlobPropagatorError extends ErrorException {
  constructor(message: string, cause: ErrorCause) {
    super(`Blob propagator failed: ${message}`, cause);
  }
}

export class BlobPropagatorCreationError extends BlobPropagatorError {
  constructor(cause: ErrorCause) {
    super("Failed to create blob propagator", cause);
  }
}
