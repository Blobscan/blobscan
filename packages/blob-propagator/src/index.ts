export type { EnvVars as Environment } from "./env";

export { env } from "./env";
export { BlobPropagator } from "./BlobPropagator";
export { blobPropagator } from "./blob-propagator";
export * from "./types";
export {
  createBlobPropagationFlowJob,
  createBlobStorageJob,
  FINALIZER_WORKER_NAME,
  STORAGE_WORKER_NAMES,
  buildJobId,
} from "./utils";
