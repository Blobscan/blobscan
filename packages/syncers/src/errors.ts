export class ErrorException extends Error {
  constructor(message: string, cause?: unknown) {
    super(message, {
      cause,
    });

    this.name = this.constructor.name;
  }
}

export class SyncerError extends ErrorException {
  constructor(syncerName: string, message: string, cause: unknown) {
    super(`Syncer "${syncerName}" failed: ${message}`, cause);
  }
}
