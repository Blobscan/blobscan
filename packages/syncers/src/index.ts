export { BaseSyncer } from "./BaseSyncer";
export { createRedisConnection } from "./utils";

export { syncSwarmStamp } from "./syncers/SwarmStampSyncer";
export { aggregateDailyStats } from "./syncers/DailyStatsSyncer";
export { aggregateOverallStats } from "./syncers/OverallStatsSyncer";
