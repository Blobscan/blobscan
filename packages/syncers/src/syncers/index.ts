export { DailyStatsSyncer } from "./DailyStatsSyncer";
export type { DailyStatsSyncerConfig } from "./DailyStatsSyncer";
export { OverallStatsSyncer } from "./OverallStatsSyncer";
export type { OverallStatsSyncerConfig } from "./OverallStatsSyncer";
