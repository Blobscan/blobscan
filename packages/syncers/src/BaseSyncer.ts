import { Queue, Worker } from "bullmq";
import type { Redis } from "ioredis";

import { createModuleLogger } from "@blobscan/logger";
import type { Logger } from "@blobscan/logger";

import { ErrorException, SyncerError } from "./errors";

export type BaseSyncerConfig = {
  name: string;
  connection: Redis;
  cronPattern: string;
  syncerFn: () => Promise<void>;
};

export class BaseSyncer {
  private name: string;
  private cronPattern: string;
  private logger: Logger;

  public readonly worker: Worker;
  public readonly queue: Queue;

  constructor({ name, cronPattern, connection, syncerFn }: BaseSyncerConfig) {
    this.name = `${name}-syncer`;
    this.cronPattern = cronPattern;
    this.logger = createModuleLogger(this.name);

    this.queue = new Queue(this.name, { connection });
    this.worker = new Worker(this.queue.name, syncerFn, { connection });

    this.queue.on("error", (error) => {
      this.logger.error(new ErrorException("A queue error occurred", error));
    });

    this.worker.on("failed", (_, error) => {
      this.logger.error(new ErrorException("A worker error occurred", error));
    });
  }

  async start() {
    try {
      const repeatableJob = await this.queue.add(`${this.name}-job`, null, {
        repeat: {
          pattern: this.cronPattern,
        },
      });

      this.logger.info("Syncer started successfully");

      return repeatableJob;
    } catch (error) {
      throw new SyncerError(
        this.name,
        "An error occurred when starting syncer",
        error
      );
    }
  }

  async close(): Promise<void> {
    const results = await Promise.allSettled([
      this.queue.obliterate({ force: true }),
      this.queue.removeAllListeners().close(),
      this.worker.removeAllListeners().close(true),
    ]);

    for (const result of results) {
      if (result.status === "rejected") {
        throw new SyncerError(
          this.name,
          "An error occurred when performing closing operation",
          result.reason
        );
      }
    }

    this.logger.info("Syncer closed successfully");
  }
}
