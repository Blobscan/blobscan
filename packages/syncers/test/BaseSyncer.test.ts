import { beforeEach, describe, expect, it, vi } from "vitest";

import { testValidError } from "@blobscan/test";

import { createRedisConnection } from "../src";
import { BaseSyncer } from "../src/BaseSyncer";
import { SyncerError } from "../src/errors";

describe("BaseSyncer", () => {
  let syncer: BaseSyncer;

  beforeEach(() => {
    syncer = new BaseSyncer({
      name: "test-updater",
      connection: createRedisConnection("redis://localhost:6379/1"),
      cronPattern: "* * * * *",
      syncerFn: async () => {
        return;
      },
    });

    return async () => syncer.close();
  });

  it("should create an updater correctly", async () => {
    const isPaused = await syncer.queue.isPaused();

    expect(
      syncer.worker.isRunning(),
      "Expected worker to be running"
    ).toBeTruthy();
    expect(isPaused, "Expected queue to be running").toBeFalsy();
  });

  describe("when running an updater", () => {
    it("should set up a repeatable job correctly", async () => {
      await syncer.start();

      const jobs = await syncer.queue.getRepeatableJobs();

      expect(jobs.length, "Expected one repeatable job").toBe(1);
      expect(jobs[0]?.pattern, "Repetable job cron pattern mismatch").toEqual(
        "* * * * *"
      );
    });

    testValidError(
      "should throw a valid error when failing to run",
      async () => {
        vi.spyOn(syncer.queue, "add").mockRejectedValueOnce(
          new Error("Queue error")
        );

        await syncer.start();
      },
      SyncerError,
      { checkCause: true }
    );
  });

  describe("when closing an updater", () => {
    it("should close correctly", async () => {
      await syncer.start();

      const queue = syncer.queue;
      const worker = syncer.worker;

      const queueCloseSpy = vi.spyOn(queue, "close").mockResolvedValueOnce();
      const queueRemoveAllListenersSpy = vi
        .spyOn(queue, "removeAllListeners")
        .mockReturnValueOnce(queue);

      const workerCloseSpy = vi.spyOn(worker, "close").mockResolvedValueOnce();
      const workerRemoveAllListenersSpy = vi
        .spyOn(worker, "removeAllListeners")
        .mockReturnValueOnce(worker);

      await syncer.close();

      expect(queueCloseSpy).toHaveBeenCalledOnce();
      expect(workerCloseSpy).toHaveBeenCalledOnce();

      expect(queueRemoveAllListenersSpy).toHaveBeenCalledOnce();
      expect(workerRemoveAllListenersSpy).toHaveBeenCalledOnce();
    });
  });

  testValidError(
    "should throw a valid error when failing to close it",
    async () => {
      vi.spyOn(syncer.queue, "close").mockRejectedValueOnce(
        new Error("Queue closing error")
      );
      vi.spyOn(syncer.worker, "close").mockRejectedValueOnce(
        new Error("Worker closing error")
      );

      await syncer.close();
    },
    SyncerError,
    {
      checkCause: true,
    }
  );
});
