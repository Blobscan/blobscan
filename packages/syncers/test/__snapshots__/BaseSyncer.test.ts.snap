// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseSyncer > should throw a valid error when failing to close it 1`] = `"Syncer \\"test-updater-syncer\\" failed: An error occurred when performing closing operation"`;

exports[`BaseSyncer > should throw a valid error when failing to close it 2`] = `[Error: Queue closing error]`;

exports[`BaseSyncer > when running an updater > should throw a valid error when failing to run 1`] = `"Syncer \\"test-updater-syncer\\" failed: An error occurred when starting syncer"`;

exports[`BaseSyncer > when running an updater > should throw a valid error when failing to run 2`] = `[Error: Queue error]`;

exports[`PeriodicUpdater > should throw a valid error when failing to close it 1`] = `"Syncer \\"test-updater-syncer\\" failed: An error occurred when performing closing operation"`;

exports[`PeriodicUpdater > should throw a valid error when failing to close it 2`] = `[Error: Queue closing error]`;

exports[`PeriodicUpdater > when running an updater > should throw a valid error when failing to run 1`] = `"Syncer \\"test-updater-syncer\\" failed: An error occurred when starting syncer"`;

exports[`PeriodicUpdater > when running an updater > should throw a valid error when failing to run 2`] = `[Error: Queue error]`;
